File: android/mekuri/src/main/kotlin/studio/anicanon/mekuri/MekuriHaptic.kt
package studio.anicanon.mekuri

import android.os.Build
import android.view.HapticFeedbackConstants

/**
 * A tactile cue in a turn: the page lifting, a drag crossing the progress at
 * which its release completes the turn, and the page landing.
 */
internal enum class MekuriHaptic {
    Lift,
    Detent,
    Land,
    ;

    /** The view feedback constant played for this cue. */
    val feedbackConstant: Int
        get() = when (this) {
            Lift -> HapticFeedbackConstants.GESTURE_START
            Detent -> if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
                HapticFeedbackConstants.SEGMENT_TICK
            } else {
                HapticFeedbackConstants.CLOCK_TICK
            }
            Land -> HapticFeedbackConstants.VIRTUAL_KEY
        }

    internal companion object {
        /**
         * Whether a drag from [from] to [to] crosses [threshold] in either
         * direction, with the inclusive comparison a release resolves by.
         */
        fun crossesThreshold(from: Float, to: Float, threshold: Float): Boolean =
            (from >= threshold) != (to >= threshold)
    }
}

File: ios/Sources/Mekuri/MekuriHaptic.swift
import SwiftUI

/// A tactile cue in a turn: the page lifting, a drag crossing the progress at
/// which its release completes the turn, and the page landing.
enum MekuriHaptic: Equatable, Sendable {
    case lift
    case detent
    case land

    /// Whether a drag from `old` to `new` crosses `threshold` in either
    /// direction, with the inclusive comparison a release resolves by.
    static func crossesThreshold(from old: CGFloat, to new: CGFloat, threshold: CGFloat) -> Bool {
        (old >= threshold) != (new >= threshold)
    }

    var feedback: SensoryFeedback {
        switch self {
        case .lift: .impact(flexibility: .soft, intensity: 0.5)
        case .detent: .selection
        case .land: .impact(flexibility: .soft, intensity: 0.9)
        }
    }
}

/// One played cue; the id makes a repeat of the same cue a new trigger.
struct MekuriHapticEvent: Equatable {
    let id: Int
    let haptic: MekuriHaptic
}
